File: valid-permutations-for-di-sequence.cc
// Valid Permutations for DI Sequence
#define FOR(i, a, b) for (remove_cv<remove_reference<decltype(b)>::type>::type i = (a); i < (b); i++)
#define REP(i, n) FOR(i, 0, n)
#define ROF(i, a, b) for (remove_cv<remove_reference<decltype(b)>::type>::type i = (b); --i >= (a); )

class Solution {
  static const int P = 1000000007;
public:
  int numPermsDISequence(string S) {
    int n = S.size();
    vector<int> f(n+1);
    f[0] = 1;
    REP(i, n)
      if (S[i] == 'D')
        ROF(j, 0, i)
          (f[j] += f[j+1]) %= P;
      else {
        int s = 0;
        REP(j, i+2) {
          int g = f[j];
          f[j] = s;
          (s += g) %= P;
        }
      }
    return accumulate(f.begin(), f.end(), 0L) % P;
  }
};

File: valid-permutations-for-di-sequence.rs
// Valid Permutations for DI Sequence
impl Solution {
    pub fn num_perms_di_sequence(s: String) -> i32 {
        const MOD: i32 = 1000000007;
        let s = s.as_bytes();
        let mut dp: Vec<i32> = vec![0; s.len()+1];
        dp[0] = 1;
        for i in 0..s.len() {
            if s[i] == b'I' {
                let mut x = 0;
                let mut ul = dp[0];
                dp[0] = 0;
                for j in 1..=i+1 {
                    x += ul; if x >= MOD { x -= MOD; }
                    ul = dp[j];
                    dp[j] = x;
                }
            } else {
                let mut x = dp[i];
                for j in (0..i).rev() {
                    x += dp[j]; if x >= MOD { x -= MOD; }
                    dp[j] = x;
                }
            }
        }
        dp.iter().fold(0, |acc,x| (acc+x)%MOD)
    }
}
